README
Copyright (C) 2003-2022 Jason Woodward <woodwardj at jaos dot org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Library General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.



Table of Contents:
 1. Features
 2. Requirements
 3. Installation
 4. slapt-getrc
 5. Getting started
 6. Troubleshooting
 7. Thanks and contributions



 1. Features

    slapt-get tries to emulate Debian's apt-get as closely as possible.

    * uses native Slackware tools (installpkg, upgradepkg, and removepkg)
    * supports packages as tgz, tlz, tbz, and txz files
    * supports multiple package sources with the ability to assign priorities to each source.
    * cache data concerning packages and updates from package sources
    * supports sources from ftp, ftps, http, https, file:/// and more (libcurl)
    * resumes broken downloads and will verify package integrity with MD5 checksum
    * show packages that are available (from cached data) and installed
    * search package listing using POSIX and extended regular expressions
      by package name, location, and description
    * retrieve, install, and remove packages by name or by specific version
    * retrieve and apply updates
    * upgrade from one Slackware release to another in a few simple steps
    * show description of packages, including mirror location, version, size,
      dependencies (where available), conflicts (where available), suggestions
      (where available), description, installation status, and the package
      changelog entry (where available).
    * exclude (hold) packages from upgrades, by name or by regular expression
    * transaction engine for install, remove, and upgrades, reporting concise
      information on what is to be done, ensuring each step happens correctly
    * "compare package version" algorithm to prevent downgrades
    * recursive dependency resolution using slack-required meta-data (see FAQ)
      supporting hard, soft, and conditional dependencies
    * package conflict resolution using slack-conflicts meta-data (see FAQ)
    * package suggestion support for complimentary packages (see FAQ)
    * Package data download robustness, only writing changes if all sources
      download successfully as well as only downloading those sources that
      have changed since the last download
    * i18n support via GNU gettext with over 30 language translations
    * extremely fast and easy to script with



 2. Requirements

    slapt-get requires Slackware-9.0 or better, due to glibc and openssl
    updates.  You can also use slapt-get under Slackware 8.1 if you
    upgrade glibc-solibs and openssl from Slackware 9.0, or if slapt-get
    is built statically linked.

    The following packages are required to build slapt-get from source:
    make, gcc, curl, glibc, kernel-headers, binutils, zlib, gettext-tools,
    and openssl.

    If gpgpme is installed, slapt-get can be built with support
    for signature verification with gnupg. gpgme requires libgpg-error.

    The unit tests in t/ require check (http://check.sourceforge.net/).



 3. Installation

    You can install via meson or make (deprecated).

    * Meson

        meson setup build # or any dir of your choosing
        meson compile -C build # or `ninja -C build` for older meson releases
        meson install -C build # or `ninja -C build install`  for older meson releases

        Meson/ninja supports `DESTDIR=/path meson install -C build`

        TIP:
        slapt-get.Slackbuild will create a Slackware package.
        Run with sudo or as a privileged user.

    * make (deprecated)

        make
        make install # as privileged user

        The Makefile honors DESTDIR.

        TIP:
        `make pkg` (as privileged user) will create a Slackware package.



 4. slapt-getrc

    The slapt-getrc file consists currently of the following keys:
     SOURCE
     WORKINGDIR
     EXCLUDE

    Choose a mirror from the slackware.com mirror list at
    http://www.slackware.com/getslack/

    * SOURCE must be a path to a valid Slackware release directory,
    or a third party source (see example within rc).
    All that is required is the PACKAGES.TXT (see FAQ entry #17)
    and CHECKSUMS.md5.  Multiple sources can be defined.

      NOTE: It is not recommended to mix 3rd party package sources
            with the official on production servers.  Sometimes
            third party sources contain the same packages as the
            official mirrors. Third party packages might not be
            as rigorously tested as official packages.

    Each SOURCE can be a assigned a priority level, by appending :{PRIORITY} to
    the source URI.  For example:

      SOURCE={url}:DEFAULT    is the default, lowest priority
      SOURCE={url}:PREFERRED  assigns more weight to this source
      SOURCE={url}:OFFICIAL   even more weight, used to denote official sources
      SOURCE={url}:CUSTOM     highest priority, for your custom package source

    See the FAQ for more information.

    * WORKINGDIR will hold the package data and downloaded packages.
      This directory will be created automatically as permissions
      allow.

    * EXCLUDE is a comma separated list of packages you want excluded from
    updating. This can be the package name, or a extended POSIX regex.
    For example, to exclude Dropline, use .*-[0-9]dl$.  See regex(7).



 5. Getting started

    After installing and customizing the slapt-getrc configuration file,
    you need to run slapt-get with the --update option to download the
    package data available from all the package sources you specified within
    the slapt-getrc file.

    Once this is done, you can use the following options:

    * --upgrade to upgrade your installed packages.
    * --dist-upgrade to upgrade to a newer release (Please refer to the FAQ
      concerning dist-upgrade prior to running the command).
    * --search to locate packages of interest by the package name, disk set,
      or version.  This supports POSIX and extended regular expressions.
    * --install to install a package.  If the package is already installed it
      will be upgraded.
    * --list to see all the packages available from the package sources you
      specified within the slapt-getrc configuration file.
    * --installed to list all the packages you already have installed.
    * --remove to remove packages.
    * --show to show information about packages.
    * --clean to remove all downloaded packages from the local cache.
    * --autoclean to remove packages from local cache that are no longer
      available on the package sources specified in the slapt-getrc file.
    * --remove-obsolete along with --remove or --dist-upgrade will mark
      all non-downloadable packages for removal.




 6. Troubleshooting

   Probably some bugs.  Email bug reports to me:
    Jason Woodward <woodwardj at jaos dot org>

   or file a github issue: https://github.com/jaos/slapt-get/issues

   Make sure --update is ran once initially and the data file (package_data)
   is good.

   Some mirrors are incomplete and the downloaded file is actually HTML
   specifying the web servers error message.  MD5 checking has been
   incorporated in order to alleviate this problem.

   Please do not add redundant mirror locations to the same release.  Mirror
   fall back is not implemented (see FAQ #50).  Also, any time the source url(s)
   are changed, slapt-get --update must be run to update the local cache.

   If you are getting segfaults, add -g to CFLAGS in the Makefile, change the
   DEBUG define in src/main.h to 1, and rebuild ( make clean && make ).  Run
   in gdb and let me know via email where the fault happens.

   Run the test suite and report any failing tests to github issues.

   See the FAQ for more info.

 7. Thanks and contributions

   John Babiak <johnb at vectorlinux dot com> is the VectorLinux slapt-get
   maintainer.

   Big thanks to Dan Barber <danb at mojolin dot com> for his insight and help
   with documentation suggestions and feature requests.

   Thanks to James Simmons over at linuxpackages.net for his interest, support,
   and ideas.

   nixon <nixon at klub dot chip dot pl> and simon
   <ssiimm at poczta dot onet dot pl> authored the Polish translation.  With
   updates by Paul Blazejowski <paulb at blazebox dot homeip dot net>, and
   Piotr Simon <sajmon at tenboard dot com>.

   Arnaldo de Moraes Pereira <arnaldo at sp-dhn dot com dot br> authored the
   Brazilian Portuguese translation.  Updates from Sergio Surkamp.

   Michiel H. <myself at mhil dot net> authored the Dutch translation.

   Joran Kvalvaag <jk at nerdworks dot org> authored the Norwegian translation.

   Ruben <ruben at snpp dot com dot ar> authored the Latin American and Spanish
   translations.  Angelos <angel dot microchip at gmail dot com> updated.

   Francesco Gigli <jaramir at slack dot it> authored the Italian translation.

   Joel Bjurman <ypolt at lediga dot st> authored the Swedish translation.

   Savvas Efstratiadis <pigouinos at mailworks dot org> authored the Greek
   translation.

   Niklas Volbers <mithrandir42 at web dot de> authored the German translation.

   Oto Petrik <oto.petrik at centrum dot cz > authored the Czech translation.

   Piotr Simon <sajmon at tenboard dot com> created the HTML version of the FAQ.

   Miguel Santinho <info at Simplicidade dot com> authored the Portuguese_Portugal
   translation with updates by Sérgio Marques <smarquesp at gmail dot com> and
   Hugo Carvalho <hugokarvalho at hotmail dot com>.

   Vu Nguyen (vncasper at yahoo dot com) authored the Vietnamese translation.

   publikum at users dot sourceforge dot net authored the Slovenian translation.

   Verachten Bruno <gounthar at gmail dot com> authored the French translation.

   freerock <freerock at gmail dot com> contributed the DESTDIR support for the makefile.

   Erik Gyepes <depi at depi dot sk> contributed the Slovak translation.

   sin of the DeepStyle team <digitalhunger at gmail dot com> contributed the Russian and
   Unkrainian translations, with updates from <ru at li dot org>.

   Victor Keranov <victor_keranov at mail dot bg> contributed the Bulgarian translation.

   Kemas Antonius <kyantonius at kyantonius dot com> contributed the Indonesian translation.

   S_A <shin1 dot abe at nifty dot com> contributed the Japanese translation.

   Yang Jinsong <shannleon at gmail dot com> contributed the simplified Chinese translation.

   Bekir SONAT <corlinux at gmail dot com> contributed the Turkish translation.

   Reteaua PTM <admin at ptm dot ro> contributed the Romanian translation.

   Ken Milmore <ken dot milmore at googlemail dot com> contributed the initial
   source priority support.

   Anders F Bjorklund <afb at users dot sourceforge dot net> contributed the
   slapt_pkg_t_filelist() support.

   Vitaly Chernookiy <vitaly.v.ch at gmail dot com> contributed the initial
   cross compiling support

   Other translators credited within the translation source files.


ChangeLog
0.11.8 Dec 25, 2022
* Curl updates

0.11.7 Jul 17, 2022
* use the configuration directory (/etc/slapt-get) for the GPG
  keyring/trustdb. This will require a rerun of --add-keys after
  installation.

0.11.6 Jul 09, 2022
* src/package.c: fix recursive dependency regression from 0.11.3
* src/main.c: fix no-upgrade regression when installing a package that is installed

0.11.5 Feb 05, 2022
 * tx pull
 * Fix case-sensitive unit comparison
 * Fixed installed package size calculation for large packages

0.11.4 Nov  8, 2021
 * Fixed dependency regression for >= or <= (thanks to Dan Church)

0.11.3 Mar 14, 2021
 * refactoring and optimizing
 * update conflict and reverse dependency handling
 * misc symbol cleanup
 * update dependency handling, fix dependency condition regression
 * meson updates
 * update unit tests
 * update common directory handling
 * fix slapt_source_t_init regression
 * update doc

0.11.2 Feb 13, 2021
 * add meson build infrastructure
 * handle new aaa_libraries (renamed from aaa_elflibs) and aaa_glibc-solibs

0.11.1 Aug 04, 2019
 * skip exclusions when removing obsolete packages
 * update naming convention for data structures
 * update dependencies
 * small cleanups
 * tx pull
 * Makefile: bump to version 0.11.1
 * rename src/curl.{c,h} to src/slaptcurl.{c,h}
 * rename src/gpgme.{c,h} to src/slaptgpgme.{c,h}
 * clang-format
 * add debug cflag
 * use slapt_vector instead of other custom list types
 * make private pkg_compare available as slapt_pkg_info_t_qsort_cmp
 * remove -ansi CFLAG for unit tests
 * update unit test test_configuration for new default rc value
 * add slapt_vector_t
 * tweak INSTALL
 * skip installing static library
 * remove binutils from slack-required

0.11.0 Mar 03, 2019
 * simplify default exclude lists
 * update software.jaos.org source locations (moving package hosting to google storage)
 * updated copyright date
 * some webservers return lowercase headers, update slapt_head_mirror_data accordingly
 * remove aaa_elflibs from exclude list, add aaa_base to dist upgrade required packages
 * check if no signatures returned from gpgme_op_verify_result
 * add slarm64 support
 * fix slapt_is_excluded check
 * refactoring, reformating, documentation updates
 * translation updates
 * openssl 1.1.x compat
 * return non-zero if --add-keys fails
 * use c99 stdbool
 * support package data in /var/lib/pkgtools
 * move default cache directory to /var/cache/slapt-get

0.10.2t Jul 04, 2016
 * add jaos.org slackwarearm source
 * 14.2 support
 * translation updates
 * fall back to provided name if parsing version out fails to find a match with show package
 * fix typo s/SLAPT_NO_SSL_VERIFYPERR/SLAPT_NO_SSL_VERIFYPEER/

0.10.2s Dec  20, 2015
 * update essential package list for dist-upgrade
 * add SLAPT_NO_SSL_VERIFYPERR environment check

0.10.2r Jul  05, 2014
 * fix conflict/reverse dep resolution when installing an alternative

0.10.2q Apr  05, 2014
 * translation updates
 * support fractional installed size from /var/log/packages files
 * --show now returns non-zero if nothing found
 * tweak search to return exact name matches in case of regex characters

0.10.2p Apr  29, 2013
 * Add newline to the end of filelist output
 * do not require root privs for filelist option (thanks David Unric)
 * Added he and pt translations from transifex
 * updated es_ES translation from Angelos

0.10.2o Aug  27, 2012
 * upgrade glibc-solibs before xz with --dist-upgrade
 * updated es_AR and es_ES translations from Fernando

0.10.2n June  25, 2012
 * remove __inline attributes for various functions as they clash with c99
 * Update pasture URL fixup for no sub directory packages
 * updated zlib usage to fix "expected 'gzFile' but argument is of type
   'struct gzFile_s **'".  Thanks Igor.

0.10.2m Jan  08, 2012
 * prefer our architecture in x86/x86_64 multilib package comparison
 * strip whitespace from url before parsing from slapt-getrc

0.10.2l Oct  08, 2011
 * added dep, conflicts, and reverse dep unit tests
 * Do not return 1 from slapt_get_pkg_dependencies if no dependency information
 * fixed reverse dep lookup to make sure the match is exact
 * only create directory structure if chdir succeeded
 * remove inclusion of deprecated curl/types.h
 * explicitly link in libcrypto
 * bumped version to 0.10.2l
 * Fix resource leaks on error paths
 * Fix typo

0.10.2k Sep  04, 2011
 * tighten up the package name regexen
 * updated translation building, russian translation from Igor

0.10.2j Jul  15, 2011
 * updated --help output (inspired by Igor's slapt-src suggestions)
 * make sure to remove the conflict's dependencies as well as the conflict
 * Updated regex for reading /var/log/packages compressed and uncompressed sizes
 * free regex and installed package list in slapt_pkg_action_filelist
 * updating unit tests
 * updated italian translation from Stefano Ansaloni
 * moved --version logout prior to reading config file
 * fixed commented out extra etc examples for 13.37
 * remove specific version in slack-desc
 * silence chdir return warning

0.10.2i Apr  27, 2011
 * set default example sources to 13.37
 * update working direction creation to account for N path depth
 * libassuan is now required via libgpgme
 * updated it translation from Stefano Ansaloni

0.10.2h Feb  15, 2011
 * fix parse_meta_tag() to match '+' in package name
 * hungarian translation by Peter Polonkai
 * updated ru translation from Igor Murzov
 * updated es_ES translation from Angelos

0.10.2g Sept 27, 2010
 * api cleanup, typedef'd old structs
 * updated el translation from George Vlahavas
 * updated pt_TR translation from Sérgio Surkam
 * updated fr translation from Verachten Bruno
 * updated es translation, new es_AR translation from Fernando Velo 
 * added slapt_download_data to public API
 * translation changes by Igor Murzov

0.10.2f June 21, 2010
 * updated es translation, new es_AR translation from Fernando Velo 
 * updated pt_BR translation from Sérgio Surkamp

src/package.h
/*
 * Copyright (C) 2003-2022 Jason Woodward <woodwardj at jaos dot org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#define SLAPT_PKG_PARSE_REGEX "(.*{1,})\\-(.*[\\-].*[\\-].*)\\.[ti]+[gblxkz]+[ ]{0,}$"
#define SLAPT_PKG_NAMEVER "(.*{1,})\\-(.*[\\-].*[\\-].*)"
#define SLAPT_PKG_VER "(.*)[\\-](.*)[\\-](.*)"
#define SLAPT_PKG_NAME_PATTERN "^PACKAGE NAME:[ ]{1,}(.*{1,})\\-(.*[\\-].*[\\-].*)(\\.[ti]+[gblzkx]+)[ ]{0,}$"
#define SLAPT_PKG_MIRROR_PATTERN "^PACKAGE MIRROR:[ ]+(.*)$"
#define SLAPT_PKG_PRIORITY_PATTERN "^PACKAGE PRIORITY:[ ]+([0-9]{1,})$"
#define SLAPT_PKG_LOCATION_PATTERN "^PACKAGE LOCATION:[ ]+(.*)$"
#define SLAPT_PKG_SIZEC_PATTERN "^PACKAGE SIZE [(]+compressed[)]{1,}:[ ]{1,}([0-9]{1,}) K$"
#define SLAPT_PKG_SIZEU_PATTERN "^PACKAGE SIZE [(]+uncompressed[)]{1,}:[ ]{1,}([0-9]{1,}) K$"
#define SLAPT_PKG_LOG_SIZEC_PATTERN "^COMPRESSED PACKAGE SIZE:[ ]{1,}([0-9\\.]{1,})[ ]{0,}([GgMmKk])$"
#define SLAPT_PKG_LOG_SIZEU_PATTERN "^UNCOMPRESSED PACKAGE SIZE:[ ]{1,}([0-9\\.]{1,})[ ]{0,}([GgMmKk])$"
#define SLAPT_PKG_LOG_DIR "/var/log/packages"
#define SLAPT_PKG_LIB_DIR "/var/lib/pkgtools/packages"
#define SLAPT_ROOT_ENV_NAME "ROOT"
#define SLAPT_ROOT_ENV_LEN 255
#define SLAPT_PKG_LOG_PATTERN "^(.*{1,})\\-(.*[\\-].*[\\-].*)"
#define SLAPT_MD5SUM_REGEX "([a-zA-Z0-9]{1,})[ ]{1,}([a-zA-Z0-9\\/._+\\-]{1,})\\/(.*{1,})\\-(.*[\\-].*[\\-].*)\\.[ti]+[gblzkx]+$"
#define SLAPT_REQUIRED_REGEX "^[ ]{0,}([^ \\<\\=\\>]{1,})[ ]{0,}([\\<\\=\\>]+){0,}[ ]{0,}([a-zA-Z0-9\\+\\.\\_\\-]+){0,}[ ]{0,}$"
#define SLAPT_MD5_STR_LEN 32
#define SLAPT_PKG_LIST "PACKAGES.TXT"
#define SLAPT_PKG_LIST_GZ "PACKAGES.TXT.gz"
#define SLAPT_PKG_LIST_L "package_data"
#define SLAPT_PATCHES_LIST "patches/PACKAGES.TXT"
#define SLAPT_PATCHES_LIST_GZ "patches/PACKAGES.TXT.gz"
#define SLAPT_CHANGELOG_FILE "ChangeLog.txt"
#define SLAPT_CHANGELOG_FILE_GZ "ChangeLog.txt.gz"
#define SLAPT_PATCHDIR "patches/"
#define SLAPT_REMOVE_CMD "/sbin/removepkg "
#define SLAPT_INSTALL_CMD "/sbin/installpkg "
#define SLAPT_UPGRADE_CMD "/sbin/upgradepkg --reinstall "
#define SLAPT_CHECKSUM_FILE "CHECKSUMS.md5"
#define SLAPT_CHECKSUM_FILE_GZ "CHECKSUMS.md5.gz"
#define SLAPT_HEAD_FILE_EXT ".head"
#define SLAPT_MAX_MMAP_SIZE 1024
#define SLAPT_MAX_ZLIB_BUFFER 1024

typedef enum {
    DEP_OP_ANY = 0,
    DEP_OP_OR,
    DEP_OP_EQ,
    DEP_OP_GTE,
    DEP_OP_LTE,
    DEP_OP_GT,
    DEP_OP_LT,
} slapt_dependency_op;

typedef struct {
    slapt_dependency_op op;
    union {
        struct {
            char *name;
            char *version;
        };
        struct {
            slapt_vector_t* alternatives;
        };
    };
} slapt_dependency_t;
slapt_dependency_t* slapt_dependency_t_init(void);
void slapt_dependency_t_free(slapt_dependency_t *);
slapt_dependency_t *slapt_dependency_t_parse_required(const char *);

typedef struct {
    char *name;
    char *version;
    char *mirror;
    char *location;
    char *description;
    char *required;
    char *conflicts;
    char *suggests;
    char *file_ext;
    slapt_vector_t *dependencies;
    uint32_t size_c;
    uint32_t size_u;
    uint32_t priority;
    bool installed;
    char md5[SLAPT_MD5_STR_LEN + 1];
} slapt_pkg_t;

typedef struct {
    slapt_pkg_t *installed;
    slapt_pkg_t *upgrade;
} slapt_pkg_upgrade_t;

slapt_pkg_upgrade_t *slapt_pkg_upgrade_t_init(slapt_pkg_t *, slapt_pkg_t *);
void slapt_pkg_upgrade_t_free(slapt_pkg_upgrade_t *);

typedef struct {
    char *pkg;
    char *error;
} slapt_pkg_err_t;

slapt_pkg_err_t *slapt_pkg_err_t_init(char *, char *);
void slapt_pkg_err_t_free(slapt_pkg_err_t *);

/* returns an empty package structure */
slapt_pkg_t *slapt_pkg_t_init(void);
/* frees the package structure */
void slapt_pkg_t_free(slapt_pkg_t *pkg);
/* generate a short description, returns (char *) on success or NULL on error, caller responsible for freeing the returned data */
char *slapt_pkg_t_short_description(const slapt_pkg_t *);
/* clean package name from package description */
char *slapt_pkg_t_clean_description(const slapt_pkg_t *);
/* generate the download url for a package, caller responsible for freeing the returned data */
char *slapt_pkg_t_url(const slapt_pkg_t *pkg);
/* compare package versions,
  returns just like strcmp,
    > 0 if a is greater than b
    < 0 if a is less than b
    0 if a and b are equal
*/
int slapt_pkg_t_cmp_versions(const char *a, const char *b);
int slapt_pkg_t_cmp(const slapt_pkg_t *a, const slapt_pkg_t *b);
/* make a copy of a package (needs to be freed with free_pkg) */
slapt_pkg_t *slapt_pkg_t_copy(slapt_pkg_t *dst, const slapt_pkg_t *src);
/* retrieve the packages changelog entry, if any.  Returns NULL otherwise, Must be chdir'd to working_dir.  */
char *slapt_pkg_t_changelog(const slapt_pkg_t *pkg);
/* returns a string representation of the package */
char *slapt_pkg_t_string(const slapt_pkg_t *pkg);
/*
  get the package filelist, returns (char *) on success or NULL on error
  caller responsible for freeing the returned data
*/
char *slapt_pkg_t_filelist(const slapt_pkg_t *pkg);
/* used by qsort */
int slapt_pkg_t_qsort_cmp(const void *a, const void *b);

/*
  update the local package cache. Must be chdir'd to working_dir.
*/
int slapt_update_pkg_cache(const slapt_config_t *global_config);
/* write pkg data to disk */
void slapt_write_pkg_data(const char *source_url, FILE *d_file, const slapt_vector_t *pkgs);
/* parse the PACKAGES.TXT file */
slapt_vector_t *slapt_parse_packages_txt(FILE *);
/*
  return a list of available packages.  Must be chdir'd to
  rc_config->working_dir.  Otherwise, open a filehandle to the package data
  and pass it to slapt_parse_packages_txt();
*/
slapt_vector_t *slapt_get_available_pkgs(void);
/* retrieve list of installed pkgs */
slapt_vector_t *slapt_get_installed_pkgs(void);

/* retrieve the newest package from package list */
slapt_pkg_t *slapt_get_newest_pkg(const slapt_vector_t *, const char *);
/* get the exact package */
slapt_pkg_t *slapt_get_exact_pkg(const slapt_vector_t *list, const char *name, const char *version);
/* lookup package by details */
slapt_pkg_t *slapt_get_pkg_by_details(const slapt_vector_t *list, const char *name, const char *version, const char *location);
/* search package list with pattern */
slapt_vector_t *slapt_search_pkg_list(const slapt_vector_t *list, const char *pattern);

/* install package by calling installpkg, returns 0 on success, -1 on error */
int slapt_install_pkg(const slapt_config_t *, const slapt_pkg_t *);
/* upgrade package by calling upgradepkg, returns 0 on success, -1 on error */
int slapt_upgrade_pkg(const slapt_config_t *global_config, const slapt_pkg_t *pkg);
/* remove package by calling removepkg, returns 0 on success, -1 on error */
int slapt_remove_pkg(const slapt_config_t *, const slapt_pkg_t *);

/* get a list of obsolete packages */
slapt_vector_t *slapt_get_obsolete_pkgs(const slapt_config_t *global_config, const slapt_vector_t *avail_pkgs, const slapt_vector_t *installed_pkgs);

/* generate the filename from the url, caller responsible for freeing the returned data */
char *slapt_gen_filename_from_url(const char *url, const char *file);
/* generate the package file name, caller responsible for freeing the returned data */
char *slapt_gen_pkg_file_name(const slapt_config_t *global_config, const slapt_pkg_t *pkg);
/* exclude pkg based on pkg name, returns 1 if package is present in the exclude list, 0 if not present */
bool slapt_is_excluded(const slapt_config_t *, const slapt_pkg_t *);
/* package is already downloaded and cached, md5sum if applicable is ok, returns slapt_code_t.  */
slapt_code_t slapt_verify_downloaded_pkg(const slapt_config_t *global_config, const slapt_pkg_t *pkg);
/* fill in the md5sum of the package */
void slapt_get_md5sums(slapt_vector_t *pkgs, FILE *checksum_file);
/* find out the pkg file size (post download) */
size_t slapt_get_pkg_file_size(const slapt_config_t *global_config, const slapt_pkg_t *pkg);

/*
  resolve dependencies
  returns 0 on success, -1 on error setting conflict_err and missing_err
  (usually called with transaction->conflict_err and transaction->missing_err)
*/
int slapt_get_pkg_dependencies(const slapt_config_t *global_config,
                               const slapt_vector_t *avail_pkgs,
                               const slapt_vector_t *installed_pkgs,
                               const slapt_pkg_t *pkg,
                               slapt_vector_t *deps,
                               slapt_vector_t *conflict_err,
                               slapt_vector_t *missing_err);
/* return list of package conflicts */
slapt_vector_t *slapt_get_pkg_conflicts(const slapt_vector_t *avail_pkgs, const slapt_vector_t *installed_pkgs, const slapt_pkg_t *pkg);
/* return list of packages required by */
slapt_vector_t *slapt_is_required_by(const slapt_config_t *global_config,
                                     const slapt_vector_t *avail,
                                     const slapt_vector_t *installed_pkgs,
                                     slapt_vector_t *pkgs_to_install,
                                     slapt_vector_t *pkgs_to_remove,
                                     const slapt_pkg_t *pkg);

/*
  clean out old outdated packages in the cache that are no longer available
  in the current source lists (ie are not downloadable)
*/
void slapt_purge_old_cached_pkgs(const slapt_config_t *global_config, const char *dir_name, slapt_vector_t *avail_pkgs);

/*
  download the PACKAGES.TXT and CHECKSUMS.md5 files
  compressed is set if the compressed version was downloaded.
*/
slapt_vector_t *slapt_get_pkg_source_packages(const slapt_config_t *global_config, const char *url, bool *compressed);
slapt_vector_t *slapt_get_pkg_source_patches(const slapt_config_t *global_config, const char *url, bool *compressed);
FILE *slapt_get_pkg_source_checksums(const slapt_config_t *global_config, const char *url, bool *compressed);
bool slapt_get_pkg_source_changelog(const slapt_config_t *global_config, const char *url, bool *compressed);
